### DISCUSSION TITLE
CLI improving

### DISCUSSION THREAD

**Discussion by pleshevskiy** [terrastruct/d2#275] [Ideas]
At the moment, for every new file we develop, we have to create a new command for build and watch. Now I achieve this with a powerful makefile setup.

```makefile
D2 := nix run ~/repos/mynix/tools\#d2_unstable --

SRC_DIR := src
OUT_DIR := out

DIAGRAMS := \
	foo \
	bar \
	biz_baz

# Supported layouts
# - dagre
# - elk
foo_theme_args := --layout dagre --theme 101
biz_baz_theme_args := --layout elk --theme 0

build: $(DIAGRAMS)

define d2rule
$(1)_args := $$(D2) $$($(1)_theme_args) $(SRC_DIR)/$(1).d2 $(OUT_DIR)/$(1).svg

$(1): prepare ; $$($(1)_args)
$(1)-w: prepare ; $$($(1)_args) -w
endef

$(foreach d2name,$(DIAGRAMS),$(eval $(call d2rule,$(d2name))))

prepare:
	mkdir -p $(OUT_DIR)

clean:
	rm -rf $(OUT_DIR)
```

And I can use it as the following:

```sh
make foo # only build
make bar-w # watch file
make biz_baz-w 
```

The idea I propose is to add an optional parameter to `--watch` on the directory where the files are located. And the d2 tool will create a small file server where it will display the current d2 files with a link to them. Once selected it will display the SVG with livereload.

---

The second suggestion is to remove `layout` and `theme` from the arguments in the d2 tool and add them to root as `direction` (if possible of course).

**Comment by ebrannin-bw** [comment]
> The second suggestion is to remove `layout` and `theme` from the arguments in the d2 tool and add them to root as `direction` (if possible of course).

For future searchers, this idea has its own issue now: #410 (though I don't think they should be _removed_ from arguments; CLI arguments should override code metadata)

**Reply by pleshevskiy** [reply]
Personally for me it became not convenient to switch `layout` to try what it looks like there. You have to turn `--watch` off, add / change CLI parameter, turn `--watch` back on. If the previous option was better, you need to repeat the steps again! This has become a daily pain for me) My opinion is that it only makes sense to leave as CLI parameters for backward compatibility

**Reply by pleshevskiy** [reply]
For clarity, I give an example that I suggest.

```d2
layout: dagre # / elk
theme: 101
direction: right

x -> y
```

instead of 

```sh
d2 --layout dagre --theme 101
```

The main point is that all the settings are stored directly in the diagram and we don't have to write a makefile to remember "what layout do I need to build this?"

**Reply by ebrannin-bw** [reply]
To be clear:
- I'm not saying your idea of `layout: elk` (etc.) in the `.d2` file is a bad idea.  I think it's a great idea!
- I _am_ saying there can be valid use-cases for keeping the CLI args -- not as a required way to change these values, but leaving them in as a higher-priority override for the optional in-diagram values.

(I'm also not part of the d2 team, just weighing in)

Edit to add: `pad` should be in this category now too (#431), though it should probably be a `style` attribute as well as an entire-diagram default-changer.

**Reply by pleshevskiy** [reply]
> I am saying there can be valid use-cases for keeping the CLI args -- not as a required way to change these values, but leaving them in as a higher-priority override for the optional in-diagram values.

Do you think so? I made a proof-of-concept wrapper today and thought the cli arguments had a lower priority than those defined in the file. I suggest you give it a try.

https://github.com/pleshevskiy/wd2

**Comment by ebrannin-bw** [comment]
Most of the discussion so far has been about setting CLI flags like `layout`, `theme`, `pad` in the `whatever.d2` source -- but most of the description is about adding a file-browser to `--watch` if you give it a directory.

~Maybe these should be separate issues?  If so, suggestions:~
- ~Change this issue's title to `Set global graph attributes in source files, not just CLI`~ I just noticed this is a Discussion, and the issue exists as #410
- Make an issue for `Support watching multiple files in a directory with --watch`? (There doesn't seem to be an issue for that yet)
